### DISCUSSION TITLE
How to translate GLSL buffer_reference with header, unsized array member and specified alignment to slang?

### DISCUSSION THREAD

**Discussion by tkerscher** [shader-slang/slang#5705] [Q&A] [answered]
I'm currently looking into switching an existing project from GLSL to slang. In it we have code that looks like the following:

```
layout(buffer_reference, scalar, buffer_reference_align=4) readonly buffer Table {
    uint n;
    float values[];
};
```

From the documentation I know that with slang I can use a struct plus pointers. However, it appears that in all of the examples the pointers point to the start of an array with elements of fixed size. Can slang do things like the example above, i.e. trailling unsized arrays? Or do I have to do some pointer magic, e.g. make the unsized a single element and take the adress of it via the ampersand operator or adding an offset to the orignal pointer? Also, how would I specify the alignment or does slang only support the default 16 byte alignment?

I hope this is the right place to ask these kind of questions and sorry if parts of it are actually related to HLSL since I'm not that proficient with it. Slang looks like a great way to get rid of the macro logic we currently have to use and it would be nice if we can test it without needing to change the code on the host side.

**Comment by expipiplus1** [comment, marked as answer]
Slang does support trailing unsized arrays.

At the moment we don't have an attribute for specifying struct alignment (however we do output alignment specification on the backend for buffer pointer types). The best design for specifying struct alignment did come up in conversation recently however, so I wouldn't be surprised if we implement something along those lines in the future.

**Comment by csyonghe** [comment]
To perform aligned load/stores, you can use `loadAligned` and `storeAligned` intrinsics: https://shader-slang.com/stdlib-reference/global-decls/loadaligned-4

**Reply by tkerscher** [reply]
I'm sorry, but that link gives me a 404 page.

**Reply by csyonghe** [reply]
I don’t know why the page got removed, but please refer to this doc instead: https://github.com/shader-slang/slang/blob/master/docs/proposals/013-aligned-load-store.md

**Comment by johannes-el** [comment]
@csyonghe @expipiplus1 Any updates on this?

How would I translate something like:
```
layout(buffer_reference, std430) readonly buffer Map {
    uint indices[];
};
```

without any alignment knowledge?

I was trying something along the lines of:
```
struct Map {
    uint indices[];
}

StructuredBuffer<Map> map;
```

But using this with:
`map = Map(x[0].data[dispatchThreadID.z].bufferAddress);`
did result in errors like the following:

```
map = Map(x[0]... expected an expression of type StructuredBuffer<Map, DefaultDataLayout>, got Map.
error: 30011: left of = is not an l-value.
```

How do I fix this and create an array of unknown size as described by OP?

**Reply by expipiplus1** [reply]
In this specific case, `StructuredBuffer` isn't RW, you want `RWStructuredBuffer` to be able to write to the elements. But there are other strange things going on in your code.

This (nonsense) shader might help you get started? It compiles to exactly the spirv you'd expect

```slang
struct Map {
    uint indices[];
}

uniform Map* map;

[shader("vertex")]
float4 vertMain() : SV_Position {
    return float4(map.indices[10]);
}
```

**Reply by johannes-el** [reply]
Hmm I don't see how that is going to help me. Maybe I should have shown a larger portion of the GLSL code:

```
layout (buffer_reference, std430) readonly buffer Max {
    uint indices[];
};

layout(local_size_x = 32, local_size_y = 32, local_size_z = 1) in;

void main() {
        Map map = Map(x.data[gl_GlobalInvocationID.z].address);
}
```

Also what does that pointer syntax in uniform do?

So in the Code I want to port there are a lot of instances of things like:
`layout(set = 0, binding = 0) readonly buffer someBuffer { vec4 normals[]; };`

I wonder how to port this, as it does require an extra struct, doesn't it?

**Reply by johannes-el** [reply]
@expipiplus1 I'd be very thankful if I could get an answer on this. Because after digging in the documentation I couldn't really find an instance in which I get a Slang equivalent to this array definition.

**Reply by expipiplus1** [reply]
@johannes-el glsl input is on a best-effort basis, and we don't currently support buffer_reference. You could open an issue requesting this feature, but I don't think that it will be a high priority for the team at the current moment. 

Do you have a minimal complete shader which demonstrates what you want to do and why it can't work with the device pointer support in Slang?

**Reply by juliusikkala** [reply]
> Also what does that pointer syntax in uniform do?

It does the same thing as buffer_reference does; a pointer. It's more that GLSL is the one with weird syntax for it here.

Instead of putting the address as a `uint64_t` in `x.data[gl_GlobalInvocationID.z].address` and constructing a `Map` with it (which is just `uint*` in disguise), that `address` could instead directly be a `uint* map_indices` and then you can just index it as you would `map.indices`. Or if you want to keep an extra `struct Map` in between, you can do what expipiplus1 suggested above.

> I wonder how to port this, as it does require an extra struct, doesn't it?

At minimum, it requires one struct less. You could just pass `vec4* normals` in push constants or that `x.data`, whatever that is in your code.

**Reply by johannes-el** [reply]
Thanks for the reply you two.
For now I don't really understand what you mean by: "That address could instead directly be a `uint*`".
Do you mean the struct member `address` should be a uint* instead of a uint64_t?

If so, I would do what?

I would do it like that?
```
struct Map {
    uint indices[];
}

uniform Map* map;

[shader("compute")]
[numthreads(32, 32, 1)]
void main(uint3 dispatchThreadID : SV_DispatchThreadID) {
    uint64_t* map_indices = x.data[dispatchThreadID.z].address;
    
    float value = 0.0;
    for (int i = 0; i < x.data[dispatchThreadID.z].size; i++) {
        uint64_t index = map_indices[i];
    }
}
```

If that is in fact correct, how do I verify that it does the same thing as the GLSL code?

So my GLSL code which is in a seperate module:

```
public struct Band {
    public uint64_t address;
    public uint64_t size;
}

public struct Uniforms {
    public Band band[MAX_FREQ];
}
```

As you might have noticed, I'm just starting with shader programming in general, so I'm not really sure what I'm doing yet. Thanks for the help thus far.

And then I use it in the code like:
`layout(set = 0, binding = 0) readonly buffer _Uniforms { Uniforms x; };`

Regarding the `vec4* normals`, I can just replace code like that:
`layout(set = 2, binding = 0) readonly buffer _SomeInstances { SomeInstanceDataGPU[] someDatas; };`

with this:
```
[[vk::push_constant]]
[[vk::binding(0, 2)]] SomeInstaceDataGPU* someDatas;
```
As you might have noticed I'm not terribly familiar with Shading languages yet, so thank you for your patience and help thus far.

**Reply by juliusikkala** [reply]
A lot of this would be easier to explain if we had more context. Like what is 'x', what the 'data' member of it is, and so on. We've only seen incomplete shards of the code, which is making it a bit difficult to give clear advice.

But yeah, what I meant is that the struct member `address` could be a `uint*` instead. The snippet you posted is a bit different, what I meant is:
```slang
[shader("compute")]
[numthreads(32, 32, 1)]
void main(uint3 dispatchThreadID : SV_DispatchThreadID) {
    uint* map_indices = x.data[dispatchThreadID.z].map_indices;
    
    float value = 0.0;
    for (int i = 0; i < x.data[dispatchThreadID.z].size; i++) {
        uint index = map_indices[i];
    }
}
```
You could verify this by simply running it and checking that it works as expected. If you are comfortable with SPIR-V, you could also compare that, although I assume you're not because you mention not being familiar with shading languages.

Just some general life advice that also applies to shading languages: Don't try to port/refactor/rewrite a whole codebase in one go, try to do it in as small pieces as possible and try running the code between each step. That may sometimes feel like extra work, but it keeps motivation up and debugging time down. You could get started by porting and compiling only one shader with Slang and keeping the rest the same as before (probably with `glslc` or `glslangValidator`).
